src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,218 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#include <stdlib.h>
-#include <stdint.h>
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
 #include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
 #include <sys/stat.h>
-#include <fcntl.h>
-#include <assert.h>
-#include <unistd.h>
 #include <time.h>
-#include <errno.h>
+#include <unistd.h>
 
 #include "pal_config.h"
 #include "pal_random.h"
+
+#if HAVE_SYS_GETRANDOM
+#include <sys/syscall.h>
+#include <sys/random.h>
+#endif // HAVE_SYS_GETRANDOM
+
 /*
 
 Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if !HAVE_ARC4RANDOM_BUF
+static int GetUrandomFileDescriptor(void)
 {
-    assert(buffer != NULL);
-
-#if HAVE_ARC4RANDOM_BUF
-    arc4random_buf(buffer, (size_t)bufferLength);
-#else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+    static int fd = -1;
 
-    if (!sMissingDevURandom)
+    if (fd < 0)
     {
-        if (rand_des == -1)
+        do
         {
-            int fd;
-
-            do
-            {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY | O_CLOEXEC);
 #else
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY);
+            fcntl(fd, F_SETFD, FD_CLOEXEC);
 #endif
-            }
-            while ((fd == -1) && (errno == EINTR));
+        }
+        while ((fd == -1) && (errno == EINTR));
+    }
 
-            if (fd != -1)
-            {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
```

Comment:
Why did this get removed? Now parallel open leaks a file descriptor.

Reply:
Not an issue any longer: the /dev/urandom is opened just once in a constructor function.  There's no need to use an atomic CAS like before.

Reply:
I see that this is now only called from a library load constructor.  So a comment explaining that would be good.